swarm/network/simulations/discovery/discovery_test.go

```diff
@@ -305,8 +307,23 @@ func triggerChecks(trigger chan discover.NodeID, net *simulations.Network, id di
 	return nil
 }
 
+// getOutboundIP gets preferred outbound ip of this machine/container
+func getOutboundIP() net.IP {
+	conn, err := net.Dial("udp", "8.8.8.8:80")
+	if err != nil {
+		panic(err)
+	}
+	defer conn.Close()
+
+	localAddr := conn.LocalAddr().(*net.UDPAddr)
+
+	return localAddr.IP
+}
+
 func newService(ctx *adapters.ServiceContext) (node.Service, error) {
-	addr := network.NewAddrFromNodeIDAndPort(ctx.Config.ID, ctx.Config.Port)
+	host := getOutboundIP()
```

Comment:
`newService` is run on specific nodes, so we get the outbound IP correctly both for `exec` and `docker` nodes. However for `exec` nodes, we no longer use localhost.